xds/src/main/java/io/grpc/xds/Bootstrapper.java

```diff
@@ -0,0 +1,162 @@
+ */
+@NotThreadSafe
+abstract class Bootstrapper {
+
+  private static final String BOOTSTRAP_PATH_SYS_ENV_VAR = "GRPC_XDS_BOOTSTRAP";
+  private static final Logger log = Logger.getLogger(Bootstrapper.class.getName());
+  private static RuntimeException failToBootstrapException;
+  private static Bootstrapper DEFAULT_INSTANCE;
+
+  static Bootstrapper getInstance() {
+    if (DEFAULT_INSTANCE == null && failToBootstrapException == null) {
+      try {
+        DEFAULT_INSTANCE = new FileBasedBootstrapper();
+      } catch (RuntimeException e) {
+        failToBootstrapException = e;
+      }
+    }
+    if (DEFAULT_INSTANCE == null) {
+      throw failToBootstrapException;
+    }
+    return DEFAULT_INSTANCE;
+  }
+
+  @Nullable
+  private static ConfigReader createConfigReaderOrNull() {
+    String filePath = System.getenv(BOOTSTRAP_PATH_SYS_ENV_VAR);
+    ConfigReader reader = null;
+    if (filePath != null) {
+      try {
+        reader = new ConfigReader(Paths.get(filePath));
+      } catch (Exception e) {
+        log.warning("Unable to read bootstrap file: " + e.getMessage());
+      }
+    } else {
+      log.warning("Environment variable " + BOOTSTRAP_PATH_SYS_ENV_VAR + " not found.");
+    }
+    return reader;
+  }
+
+  /**
+   * Returns the canonical name of the traffic director to be connected to.
+   */
+  abstract String getBalancerName();
+
+  /**
+   * Returns a {@link Node} message with project/network metadata in it to be included in
+   * xDS requests.
+   */
+  abstract Node getNode();
+
+  /**
+   * Returns the credentials to use when communicating with the xDS server.
+   */
+  abstract CallCredentials getCallCredentials();
+
+  @VisibleForTesting
+  static final class FileBasedBootstrapper extends Bootstrapper {
+
+    private final String balancerName;
+    private final Node node;
+    // TODO(chengyuanzhang): Add configuration for call credentials loaded from bootstrap file.
+    //  hard-coded for alpha release.
+
+    private FileBasedBootstrapper() {
+      this(createConfigReaderOrNull());
+    }
+
+    @VisibleForTesting
+    FileBasedBootstrapper(@Nullable ConfigReader reader) {
+      if (reader == null) {
+        throw new RuntimeException("Failed to bootstrap from config file.");
+      }
+      balancerName = reader.getServerConfig().getGrpcServices(0).getGoogleGrpc().getTargetUri();
+      node = reader.getNode();
+    }
+
+    @Override
+    String getBalancerName() {
+      return balancerName;
+    }
+    
+    @Override
+    Node getNode() {
+      return node;
+    }
+
+    @Override
+    CallCredentials getCallCredentials() {
+      return MoreCallCredentials.from(ComputeEngineCredentials.create());
+    }
+  }
+
+  @VisibleForTesting
+  static final class ConfigReader {
+    private final Node node;
+    private final ApiConfigSource serverConfig;
+
+    private ConfigReader(Path path) throws IOException {
+      this(new String(Files.readAllBytes(path), Charset.defaultCharset()));
```

Comment:
Do not use defaultCharset. Use UTF-8 or ASCII instead.

I'll grant you that this could be a reasonable place to use defaultCharset, but using the default still is only likely to cause problems.